Lifecycle callbacks do not work

# Bug Report

None of the callbacks get triggered after a succesful reflex

## Expected behavior

afterReflex should trigger after every reflex 

```
afterReflex(element, reflex)  {
 console.log('callback works')
}

```
 
## Versions

### StimulusReflex

3.3.0-pre2

### External tools

- Ruby: 2.7
- Rails: 6.0.0

### Browser

- Chrome
-  84.0.4147.89

here are errors from webpack 

```
ERROR in ./app/javascript/controllers/task_activity_controller.js
Module not found: Error: Can't resolve 'stimulus_reflex' in '/mnt/c/Users/Documents/testapp/app/javascript/cont
rollers'
resolve 'stimulus_reflex' in '/mnt/c/Users/Documents/testapp/app/javascript/controllers'
  Parsed request is a module
  using description file: /mnt/c/Users/Documents/testapp/package.json (relative path: ./app/javascript/controll
ers)
```

--- a/javascript/lifecycle.js
+++ b/javascript/lifecycle.js
@@ -11,7 +11,7 @@ import { camelize } from './utils'
 //
 // - element - the element that triggered the reflex (not necessarily the Stimulus controller's element)
 //
-const invokeLifecycleMethod = (stage, element) => {
+const invokeLifecycleMethod = (stage, element, reflexId) => {
   if (!element || !element.reflexData) return
   const controller = element.reflexController
   const reflex = element.reflexData.target
@@ -33,7 +33,8 @@ const invokeLifecycleMethod = (stage, element) => {
         controller,
         element,
         reflex,
-        element.reflexError
+        element.reflexError,
+        reflexId
       )
     )
   }
@@ -44,7 +45,8 @@ const invokeLifecycleMethod = (stage, element) => {
         controller,
         element,
         reflex,
-        element.reflexError
+        element.reflexError,
+        reflexId
       )
     )
   }
@@ -59,15 +61,15 @@ const invokeLifecycleMethod = (stage, element) => {
 
 document.addEventListener(
   'stimulus-reflex:before',
-  event => invokeLifecycleMethod('before', event.target),
+  event => invokeLifecycleMethod('before', event.target, event.detail.reflexId),
   true
 )
 
 document.addEventListener(
   'stimulus-reflex:success',
   event => {
-    invokeLifecycleMethod('success', event.target)
-    dispatchLifecycleEvent('after', event.target)
+    invokeLifecycleMethod('success', event.target, event.detail.reflexId)
+    dispatchLifecycleEvent('after', event.target, event.detail.reflexId)
   },
   true
 )
@@ -75,8 +77,8 @@ document.addEventListener(
 document.addEventListener(
   'stimulus-reflex:selector',
   event => {
-    invokeLifecycleMethod('success', event.target)
-    dispatchLifecycleEvent('after', event.target)
+    invokeLifecycleMethod('success', event.target, event.detail.reflexId)
+    dispatchLifecycleEvent('after', event.target, event.detail.reflexId)
   },
   true
 )
@@ -84,8 +86,8 @@ document.addEventListener(
 document.addEventListener(
   'stimulus-reflex:nothing',
   event => {
-    invokeLifecycleMethod('success', event.target)
-    dispatchLifecycleEvent('after', event.target)
+    invokeLifecycleMethod('success', event.target, event.detail.reflexId)
+    dispatchLifecycleEvent('after', event.target, event.detail.reflexId)
   },
   true
 )
@@ -93,21 +95,21 @@ document.addEventListener(
 document.addEventListener(
   'stimulus-reflex:error',
   event => {
-    invokeLifecycleMethod('error', event.target)
-    dispatchLifecycleEvent('after', event.target)
+    invokeLifecycleMethod('error', event.target, event.detail.reflexId)
+    dispatchLifecycleEvent('after', event.target, event.detail.reflexId)
   },
   true
 )
 
 document.addEventListener(
   'stimulus-reflex:halted',
-  event => invokeLifecycleMethod('halted', event.target),
+  event => invokeLifecycleMethod('halted', event.target, event.detail.reflexId),
   true
 )
 
 document.addEventListener(
   'stimulus-reflex:after',
-  event => invokeLifecycleMethod('after', event.target),
+  event => invokeLifecycleMethod('after', event.target, event.detail.reflexId),
   true
 )
 
@@ -122,7 +124,7 @@ document.addEventListener(
 //
 // - element - the element that triggered the reflex (not necessarily the Stimulus controller's element)
 //
-export const dispatchLifecycleEvent = (stage, element) => {
+export const dispatchLifecycleEvent = (stage, element, reflexId) => {
   if (!element) return
   const { target } = element.reflexData || {}
   element.dispatchEvent(
@@ -131,7 +133,8 @@ export const dispatchLifecycleEvent = (stage, element) => {
       cancelable: false,
       detail: {
         reflex: target,
-        controller: element.reflexController
+        controller: element.reflexController,
+        reflexId
       }
     })
   )

--- a/javascript/stimulus_reflex.js
+++ b/javascript/stimulus_reflex.js
@@ -133,7 +133,7 @@ const extendStimulusController = controller => {
       element.reflexController = this
       element.reflexData = data
 
-      dispatchLifecycleEvent('before', element)
+      dispatchLifecycleEvent('before', element, reflexId)
 
       setTimeout(() => {
         const { params } = element.reflexData || {}
@@ -168,6 +168,9 @@ const extendStimulusController = controller => {
           data
         }
       })
+
+      promise.reflexId = reflexId
+
       if (debugging) promise.catch(() => {}) // noop default catch
       return promise
     },
@@ -384,7 +387,7 @@ if (!document.stimulusReflexInitialized) {
         promise.resolve(response)
       }
 
-      dispatchLifecycleEvent('success', element)
+      dispatchLifecycleEvent('success', element, reflexId)
       if (debugging) Log.success(response)
     })
   }

--- a/javascript/stimulus_reflex.js
+++ b/javascript/stimulus_reflex.js
@@ -40,7 +40,7 @@ const createSubscription = controller => {
   actionCableConsumer = actionCableConsumer || getConsumer()
   const { channel } = controller.StimulusReflex
   const identifier = JSON.stringify({ channel })
-  let totalOperations = 0
+  let totalOperations
   let reflexId
 
   controller.StimulusReflex.subscription =
@@ -48,6 +48,7 @@ const createSubscription = controller => {
     actionCableConsumer.subscriptions.create(channel, {
       received: data => {
         if (!data.cableReady) return
+        totalOperations = 0
         ;['morph', 'innerHtml'].forEach(operation => {
           if (data.operations[operation] && data.operations[operation].length) {
             if (data.operations[operation][0].stimulusReflex) {
@@ -62,7 +63,6 @@ const createSubscription = controller => {
             }
           }
         })
-
         if (promises[reflexId]) {
           promises[reflexId].totalOperations = totalOperations
           promises[reflexId].completedOperations = 0

--- a/javascript/stimulus_reflex.js
+++ b/javascript/stimulus_reflex.js
@@ -40,20 +40,34 @@ const createSubscription = controller => {
   actionCableConsumer = actionCableConsumer || getConsumer()
   const { channel } = controller.StimulusReflex
   const identifier = JSON.stringify({ channel })
+  let totalOperations = 0
+  let reflexId
 
   controller.StimulusReflex.subscription =
     actionCableConsumer.subscriptions.findAll(identifier)[0] ||
     actionCableConsumer.subscriptions.create(channel, {
       received: data => {
         if (!data.cableReady) return
-        if (data.operations.morph && data.operations.morph.length) {
-          if (data.operations.morph[0].stimulusReflex) {
-            const urls = Array.from(
-              new Set(data.operations.morph.map(m => m.stimulusReflex.url))
-            )
-            if (urls.length !== 1 || urls[0] !== location.href) return
+        ;['morph', 'innerHtml'].forEach(operation => {
+          if (data.operations[operation] && data.operations[operation].length) {
+            if (data.operations[operation][0].stimulusReflex) {
+              const urls = Array.from(
+                new Set(
+                  data.operations[operation].map(m => m.stimulusReflex.url)
+                )
+              )
+              if (urls.length !== 1 || urls[0] !== location.href) return
+              totalOperations += data.operations[operation].length
+              reflexId = data.operations[operation][0].stimulusReflex.reflexId
+            }
           }
+        })
+
+        if (promises[reflexId]) {
+          promises[reflexId].totalOperations = totalOperations
+          promises[reflexId].completedOperations = 0
         }
+
         CableReady.perform(data.operations)
       }
     })
@@ -368,29 +382,28 @@ if (!document.stimulusReflexInitialized) {
   const beforeDOMUpdateHandler = event => {
     const { selector, stimulusReflex } = event.detail || {}
     if (!stimulusReflex) return
-    const { reflexId, attrs, last } = stimulusReflex
+    const { reflexId, attrs } = stimulusReflex
     const element = findElement(attrs)
     const promise = promises[reflexId]
 
-    if (!last) return
+    promise.completedOperations++
+    if (promise.completedOperations < promise.totalOperations) return
 
     const response = {
       element,
       event,
-      morphMode: promise && promise.morphMode,
       data: promise && promise.data
     }
 
-    setTimeout(() => {
-      if (promise) {
-        delete promises[reflexId]
-        promise.resolve(response)
-      }
+    if (promise) {
+      delete promises[reflexId]
+      promise.resolve(response)
+    }
 
-      dispatchLifecycleEvent('success', element, reflexId)
-      if (debugging) Log.success(response)
-    })
+    dispatchLifecycleEvent('success', element, reflexId)
+    if (debugging) Log.success(response)
   }
+
   document.addEventListener(
     'cable-ready:before-inner-html',
     beforeDOMUpdateHandler
@@ -427,7 +440,8 @@ if (!document.stimulusReflexInitialized) {
       }
     }
 
-    if (element && subjects[subject]) dispatchLifecycleEvent(subject, element)
+    if (element && subjects[subject])
+      dispatchLifecycleEvent(subject, element, reflexId)
 
     if (debugging) {
       switch (subject) {

--- a/lib/stimulus_reflex/broadcasters/broadcaster.rb
+++ b/lib/stimulus_reflex/broadcasters/broadcaster.rb
@@ -29,6 +29,7 @@ module StimulusReflex
       cable_ready[stream_name].dispatch_event(
         name: "stimulus-reflex:server-message",
         detail: {
+          reflexId: data["reflexId"],
           stimulus_reflex: data.merge(
             broadcaster: to_sym,
             server_message: {subject: subject, body: body}

--- a/lib/stimulus_reflex/broadcasters/page_broadcaster.rb
+++ b/lib/stimulus_reflex/broadcasters/page_broadcaster.rb
@@ -18,7 +18,6 @@ module StimulusReflex
           children_only: true,
           permanent_attribute_name: permanent_attribute_name,
           stimulus_reflex: data.merge({
-            last: selector == selectors.last,
             broadast_type: to_sym
           })
         )

--- a/lib/stimulus_reflex/broadcasters/selector_broadcaster.rb
+++ b/lib/stimulus_reflex/broadcasters/selector_broadcaster.rb
@@ -7,7 +7,6 @@ module StimulusReflex
         selectors, html = morph
         updates = selectors.is_a?(Hash) ? selectors : Hash[selectors, html]
         updates.each do |selector, html|
-          last = morph == morphs.last && selector == updates.keys.last
           html = html.to_s
           fragment = Nokogiri::HTML.fragment(html)
           match = fragment.at_css(selector)
@@ -18,7 +17,6 @@ module StimulusReflex
               children_only: true,
               permanent_attribute_name: permanent_attribute_name,
               stimulus_reflex: data.merge({
-                last: last,
                 broadast_type: to_sym
               })
             )
@@ -27,7 +25,6 @@ module StimulusReflex
               selector: selector,
               html: fragment.to_html,
               stimulus_reflex: data.merge({
-                last: last,
                 broadast_type: to_sym
               })
             )

